### ISSUE TITLE
Still problem with validation and Dynamic filled dropdowns

### ISSUE THREAD

**Issue by ArfyFR** [google/blockly#3638]
The issue I opened is closed, https://github.com/google/blockly/issues/2926 , so I open a new one.
Hi, we tested with the last Blockly, 3.20191014.4, and we still have the problem with dynamic filled dropdowns.

An example to reproduce the problem with dynamic filled dropdown (in function of another dropdown value)

[blockly_dyn_fields.zip](https://github.com/google/blockly/files/4102366/blockly_dyn_fields.zip)
(I used last blockly from today)

By doing this: using a function to fill FieldDropdown (with generateOptions a function)
```
this.appendDummyInput()
        .appendField(new Blockly.FieldDropdown(generateOptions), "FIELD_NAME");
```
then this error occurs:
```
Cannot set the dropdown's value to an unavailable option. Block type: test_dyn_field_field_block_type, Field name: FIELD_NAME, Value: SET2_FIELD3
Blockly.FieldDropdown.doClassValidation_ @ field_dropdown.js:420
Blockly.Field.setValue @ field.js:726
Blockly.Field.fromXml @ field.js:353
Blockly.Xml.domToField_ @ xml.js:802
Blockly.Xml.domToBlockHeadless_ @ xml.js:695
Blockly.Xml.domToBlockHeadless_ @ xml.js:733
Blockly.Xml.domToBlockHeadless_ @ xml.js:709
Blockly.Xml.domToBlock @ xml.js:544
Blockly.Xml.domToWorkspace @ xml.js:415
(anonymous) @ index.html:33
```

**Comment by samelhusseini** [comment]
The issue here is that blocks get rendered from the bottom (deepest in a tree) upwards, so as a FieldDropdown gets initialized and rendered and ``getOptions`` is called, getRootBlock yield's the block itself and not the connected block. It's not until a user clicks on the dropdown do we have a connection to the root block and your logic works.

I tracked this back to Blockly v1.20190419.0 and the only difference is we used to let you set an invalid field option whereas now we validate the options in the FieldDropdown's constructor.

You could possibly change your logic to add an onchange that re-renders the field when a block is moved, something like this:
```

Blockly.Blocks['test_dyn_field_field_block_type'] = {
  init: function() {
    const generateOptions = (function(that) {
      return function() {
        let rootBlock = that.getRootBlock();
        if(rootBlock == null) {
          return [['NO_ROOT', 'NO_ROOT']];
        }

        if(rootBlock.type !== 'test_dyn_field_root_block_type') {
          return [['WRONG_TYPE', 'WRONG_TYPE']];
        }

        let setName = rootBlock.getFieldValue('SET_NAME');
        if(setName === "SET1_NAME") {
          return [['field1_1', 'SET1_FIELD1'], ['field1_2', 'SET1_FIELD2'], ['field1_3', 'SET1_FIELD3']];
        } else if(setName === "SET2_NAME") {
          return [['field2_1', 'SET2_FIELD1'], ['field2_2', 'SET2_FIELD2'], ['field2_3', 'SET2_FIELD3']];
        } else if(setName === "SET3_NAME") {
          return [['field3_1', 'SET3_FIELD1'], ['field3_2', 'SET3_FIELD2'], ['field3_3', 'SET3_FIELD3']];
        }

        return [['WRONG_SET', 'WRONG_SET']];
      }
    })(this);
    this.appendDummyInput()
        .appendField(new Blockly.FieldDropdown(generateOptions), "FIELD_NAME");
    this.setPreviousStatement(true, "test_dyn_field_field_block_type");
    this.setNextStatement(true, "test_dyn_field_field_block_type");
    this.setColour(230);
    this.setTooltip("");
    this.setHelpUrl("");
    this.oldRootBlock_ = this.getRootBlock();
  },
  onchange: function (e) {
    if (e.type == Blockly.Events.FINISHED_LOADING || e.type == Blockly.Events.MOVE) {
      if (this.getRootBlock() != this.oldRootBlock_) {
        var field = this.getField('FIELD_NAME');
        var options = field.getOptions(false);
        field.setValue(options[0][1]);
        this.oldRootBlock_ = this.getRootBlock();
      }
    }
  }
};
```

**Comment by ArfyFR** [comment]
Hi,
As I said in previous issue https://github.com/google/blockly/issues/2926, we save/restore workspace using
`Blockly.Xml.domToWorkspace(this.contentData.content, this.workspace);
`
and our problem is with the field.setValue at start during the domToWorkspace

User story is:

- create a valid "blockly", 
- save it as xml
- close editor
- next day come back, needs to edit the saved valid "blockly"

=> here problem due to "new validation way " for dynamic filled dropdowns

May be postpone the validation after all the data restoration ?

**Comment by rachel-fenichel** [comment]
I've added this to our milestone for the next release--not being able to load XML generated from Blockly is definitely an issue.  I don't think postponing validation until after restoration is an option for us, though.

**Comment by samelhusseini** [comment]
It's unclear to me what the next steps are here. The problem is as follows: 
The user selects a dropdown option that is valid at the time, and so the FieldDropdown saves that value, which is then serialized to XML and saved as the field value.
Later, when the XML is imported into Blockly, that option is no longer valid which doesn't allow the value to be set, and outputs a warning.

We don't provide a way to turn off that validation option (as it's a class validator). @BeksOmega if you have some ideas.

**Comment by BeksOmega** [comment]
> @BeksOmega if you have some ideas.

I don't have any good ones :/

**Option 1**
Rip up xml.js. Change it so that we connect blocks before deserializing the field values. Pretty sure no one wants to do this because it would be super dangerous.

Also this would break any developer-specified connection-availability functions you may want to add in the future. Because in that case you'd want to know the field values before connecting the blocks.

**Option 2**
Rip up xml.js. Add a function to fields that turns off class validators. Follow this process in xml.js:
1. Turn off field validation.
2. Set all field values.
3. Connect the blocks.
4. Turn on field validation.
5. Set all field values again.

Again, dangerous. Also slow when no one wants deserializing to be slow.

**Option 3**
Add documentation about the limitations of validators.

I don't know anything about this particular situation, because I haven't seen this block. But I think that generally if one block's dropdown depends on its parent in this way, they could either be combined into one, or the dropdown block could be split into more blocks.

That being the case I think it's reasonable to ask the developer to handle the case where their block isn't connected to anything. All they would have to do is return their most permissive set of options. Or have a mutator that stores the type of the parent, since mutators are (generally) deserialized before fields.

Limitations of validators (i.e. what cases your validator may have to deal with during deserialization) could be included on the [validators docs page](https://developers.google.com/blockly/guides/create-custom-blocks/fields/validators).

**Comment by ArfyFR** [comment]
Minimal for me would be
 - API to disable validation
 - API to re-enable validation

For the moment I did this in our code:


    // TODO check next Blockly version cf issue https://github.com/google/blockly/issues/2926
    // Temporary disable dropdown validation as we use dynamic one
    const fieldDropdownDoClassValidation_ = Blockly.FieldDropdown.prototype.doClassValidation_;
    Blockly.FieldDropdown.prototype.doClassValidation_ = function(newValue): string {
      return newValue;
    };

    // Load blocks
    if (this.contentData.content != null) {
      Blockly.Xml.domToWorkspace(this.contentData.content, this.workspace);
    }

    // TODO check next Blockly version cf issue https://github.com/google/blockly/issues/2926
    // Temporary disable dropdown validation as we use dynamic one
    Blockly.FieldDropdown.prototype.doClassValidation_ = fieldDropdownDoClassValidation_;

**Comment by rachel-fenichel** [comment]
The problem here is that we want a strict guarantee that the value on the block is valid at all times.  Turning off validation on all blocks breaks that.  I think it has to be on a per-block or per-field case, that the developer has to add a flag to disable validation during loading.

Would it be sufficient to not validate on your blocks during loading, and then revalidate when you receive the onFinishLoading event?

**Comment by ArfyFR** [comment]
> Would it be sufficient to not validate on your blocks during loading, and then revalidate when you receive the onFinishLoading event?

Or another "load" API, for example
`Blockly.Xml.domToWorkspace_validateOnlyAfterBlocksConnected(this.contentData.content, this.workspace);
`

You can test it with project example I provided: blockly_dyn_fields.zip in my first 
https://github.com/google/blockly/issues/3638#issue-554049786

**Comment by xylix** [comment]
We use this workaround at [funkly](https://github.com/funktionaalinen-graafinen-kieli/funkly) and it works pretty fine. Would be nice to have a toggle though.
```
const prevValidation = Blockly.FieldDropdown.prototype.doClassValidation_
// turn validations off, saving the old validation function
Blockly.FieldDropdown.prototype.doClassValidation_ = function (newValue: any) {
    return newValue
} 
// import code

// Turn validations back on
Blockly.FieldDropdown.prototype.doClassValidation_ = p
```

**Comment by BeksOmega** [comment]
Note for this: At the most recent summit there were questions about whether deserializing random XML into a headless workspace and then converting it to code was safe. And the current answer is yes (as far as we know). But if you turn off validation when loading XML, that's not necessarily the case anymore. So I think it's best if we continue validating field values when loading from XML.

If anyone is still running into this issue, I recommend going with Option 3 from my post above. When your block is in a disconnected state, have your dynamic dropdown return the most permissive set of values possible, and you should be good.

**Comment by jschanker** [comment]
@ArfyFR @BeksOmega I'm not sure what change may have corrected this, but loading from the XML seems to work correctly now without any warning about unavailable options in the Blockly Playground.  Tested as follows:

1. Go to http://google.github.io/blockly/tests/playground.html
2. Enter the following block definitions in the Developer console:
```
Blockly.Blocks['test_dyn_field_root_block_type'] = {
  init: function() {
    this.appendDummyInput()
        .appendField(new Blockly.FieldDropdown([["set1","SET1_NAME"], ["set2","SET2_NAME"], ["set3","SET3_NAME"]]), "SET_NAME");
    this.appendStatementInput("FIELDS")
        .setCheck("test_dyn_field_field_block_type");
    this.setColour(230);
    this.setTooltip("");
    this.setHelpUrl("");
  }
};

Blockly.Blocks['test_dyn_field_field_block_type'] = {
  init: function() {
    const generateOptions = (function(that) {
      return function() {
        let rootBlock = that.getRootBlock();
        if(rootBlock == null) {
          return [['NO_ROOT', 'NO_ROOT']];
        }

        if(rootBlock.type !== 'test_dyn_field_root_block_type') {
          return [['WRONG_TYPE', 'WRONG_TYPE']];
        }

        let setName = rootBlock.getFieldValue('SET_NAME');
        if(setName === "SET1_NAME") {
          return [['field1_1', 'SET1_FIELD1'], ['field1_2', 'SET1_FIELD2'], ['field1_3', 'SET1_FIELD3']];
        } else if(setName === "SET2_NAME") {
          return [['field2_1', 'SET2_FIELD1'], ['field2_2', 'SET2_FIELD2'], ['field2_3', 'SET2_FIELD3']];
        } else if(setName === "SET3_NAME") {
          return [['field3_1', 'SET3_FIELD1'], ['field3_2', 'SET3_FIELD2'], ['field3_3', 'SET3_FIELD3']];
        }

        return [['WRONG_SET', 'WRONG_SET']];
      }
    })(this);

    this.appendDummyInput()
        .appendField(new Blockly.FieldDropdown(generateOptions), "FIELD_NAME");
    this.setPreviousStatement(true, "test_dyn_field_field_block_type");
    this.setNextStatement(true, "test_dyn_field_field_block_type");
    this.setColour(230);
    this.setTooltip("");
    this.setHelpUrl("");
  },
};
```
3. Load the following XML:
```
<xml xmlns="https://developers.google.com/blockly/xml">
  <block type="test_dyn_field_root_block_type">
    <field name="SET_NAME">SET2_NAME</field>
    <statement name="FIELDS">
      <block type="test_dyn_field_field_block_type">
        <field name="FIELD_NAME">SET2_FIELD3</field>
      </block>
    </statement>
  </block>
</xml>
```

Am I correct about this being the issue that didn't work before?

**Comment by BeksOmega** [comment]
> Am I correct about this being the issue that didn't work before?

I believe so yes. I think this was fixed when an [explicit deserialization order was added](https://github.com/google/blockly/pull/4571). This change made it so that blocks are connected to their parents (but not children) before fields are deserialized. So now dynamic dropdowns can properly access the parent during deserialization.

Before, xml nodes were just processed in the order they appeared. Which mean fields were usually processed before blocks were connected to their parents. See the [previous version](https://github.com/google/blockly/blob/60e96c4d93d03c39ce680358846bba9d5d78299a/core/xml.js#L625).
